fix: correct terminal width for halfwidth katakana with dakuten

g4yamanaka opened the pull request:
## Problem

Halfwidth katakana combined with halfwidth dakuten (ﾞ U+FF9E) or handakuten (ﾟ U+FF9F) causes layout corruption in ratatui. This occurs in two specific cases:

1. **Border titles**: Text overflows the box boundary
2. **Background colors**: Background extends beyond the intended area

### Root Cause

The issue stems from a mismatch between how `unicode-width` calculates width and how terminals actually display these characters:

- **unicode-width calculation**: Halfwidth dakuten/handakuten are reported as width 0 (due to their `Grapheme_Extend` Unicode property)
- **Terminal display**: These characters are displayed as width 1 each (inherited from legacy Shift-JIS encoding where they were separate 1-byte characters)

For example, `ｶﾞｷﾞｸﾞ` consists of 3 grapheme clusters (`ｶﾞ`, `ｷﾞ`, `ｸﾞ`):
- **Correct width**: Each cluster should be 2 columns (halfwidth katakana 1 + dakuten 1), total = 6 columns
- **Ratatui's calculation**: Each cluster calculated as 1 column (dakuten reported as 0), total = 3 columns
- **Terminal rendering**: Each cluster actually takes 2 columns, total = 6 columns
- **Result**: Buffer allocated for 3 columns but content requires 6 columns → **3 columns overflow**

## Reproduction

### Visual Comparison

**Before fix (ratatui 0.29.0):**
<img width="180" height="125" alt="before" src="https://github.com/user-attachments/assets/163e2846-cb4d-42f1-affb-c4730614c458" />

**After fix:**
<img width="180" height="125" alt="after" src="https://github.com/user-attachments/assets/b67c40ec-cdaf-4512-8bc4-9905cad81ff4" />

### Reproduction Code

```rust
use crossterm::{
    event::{self, Event, KeyCode},
    execute,
    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
};
use ratatui::{
    Frame, Terminal, backend::CrosstermBackend, layout::{Constraint, Layout}, style::{Color, Style}, widgets::{Block, Borders, Paragraph}
};
use std::io;

fn main() -> io::Result<()> {
    enable_raw_mode()?;
    let mut stdout = io::stdout();
    execute!(stdout, EnterAlternateScreen)?;
    let backend = CrosstermBackend::new(stdout);
    let mut terminal = Terminal::new(backend)?;

    let res = run_app(&mut terminal);

    disable_raw_mode()?;
    execute!(terminal.backend_mut(), LeaveAlternateScreen)?;
    terminal.show_cursor()?;

    res
}

fn run_app(terminal: &mut Terminal<CrosstermBackend<io::Stdout>>) -> io::Result<()> {
    loop {
        terminal.draw(ui)?;

        if let Event::Key(key) = event::read()? {
            if key.code == KeyCode::Char('q') {
                return Ok(());
            }
        }
    }
}

fn ui(frame: &mut Frame) {
    let area = frame.area();

    // Create two narrow columns (10 columns each) side by side
    let columns = Layout::default()
        .direction(ratatui::layout::Direction::Horizontal)
        .constraints([
            Constraint::Length(10),
            Constraint::Length(10),
        ])
        .split(area);

    // Left column: halfwidth katakana + dakuten (buggy)
    let left_chunks = Layout::default()
        .direction(ratatui::layout::Direction::Vertical)
        .constraints([
            Constraint::Length(3),
            Constraint::Length(3),
        ])
        .split(columns[0]);

    // Right column: ASCII alphabet (reference, works correctly)
    let right_chunks = Layout::default()
        .direction(ratatui::layout::Direction::Vertical)
        .constraints([
            Constraint::Length(3),
            Constraint::Length(3),
        ])
        .split(columns[1]);

    // Case 1: Border title with "ｶﾞｷﾞ" (2 grapheme clusters)
    // Each cluster (ｶﾞ, ｷﾞ) should be width 2, total should be 4
    // But ratatui calculates as width 2 (1 per cluster)
    // Terminal displays as width 4 → overflows by 2 columns
    frame.render_widget(
        Paragraph::new("Case 1")
            .block(Block::default()
                .borders(Borders::ALL)
                .title("ｶﾞｷﾞ")),
        left_chunks[0],
    );

    // Case 2: Content with "ｶﾞｷﾞｸﾞ" (3 grapheme clusters)
    // Each cluster (ｶﾞ, ｷﾞ, ｸﾞ) should be width 2, total should be 6
    // But ratatui calculates as width 3 (1 per cluster)
    // Terminal displays as width 6 → overflows by 3 columns
    frame.render_widget(
        Paragraph::new("ｶﾞｷﾞｸﾞ")
            .style(Style::default().bg(Color::Red))
            .block(Block::default()
                .borders(Borders::ALL)
                .title("Case 2")),
        left_chunks[1],
    );

    // Reference boxes with ASCII (no overflow)
    frame.render_widget(
        Paragraph::new("AB")
            .block(Block::default()
                .borders(Borders::ALL)
                .title("AB")),
        right_chunks[0],
    );

    frame.render_widget(
        Paragraph::new("ABC")
            .style(Style::default().bg(Color::Green))
            .block(Block::default()
                .borders(Borders::ALL)
                .title("ABC")),
        right_chunks[1],
    );
}
```

Run with `ratatui = "0.29.0"` to see the overflow issue.

## Solution

Introduced a new `TerminalWidthStr` trait that correctly calculates the display width for grapheme clusters containing halfwidth dakuten/handakuten.

### Implementation Details

1. **Added `TerminalWidthStr` trait**
   - New trait in `ratatui-core/src/text/grapheme.rs`
   - Provides `.terminal_width()` method for `str` types
   - Correctly calculates display width for grapheme clusters with halfwidth dakuten/handakuten

2. **Internal implementation**
   - Internal `terminal_width(grapheme: &str) -> usize` function
   - When a grapheme cluster ends with halfwidth dakuten/handakuten (U+FF9E or U+FF9F), calculates width by summing individual character widths, treating dakuten/handakuten as width 1
   - For all other cases, uses standard `unicode-width` calculation

3. **Updated all width calculations across the codebase**
   - All width calculations now use `.terminal_width()` method:
     - `ratatui-core/src/backend/test.rs`
     - `ratatui-core/src/buffer/buffer.rs`
     - `ratatui-core/src/text/span.rs`
     - `ratatui-widgets/src/barchart/bar.rs`
     - `ratatui-widgets/src/reflow.rs`
     - `ratatui-widgets/src/scrollbar.rs`
     - `ratatui-widgets/src/tabs.rs`

## Testing

Added the following tests:

- Tests for halfwidth dakuten/handakuten alone
- Tests for valid combinations with halfwidth katakana
- Tests for non-katakana combinations
- Tests for combining dakuten/handakuten (U+3099, U+309A)
- Tests for `TerminalWidthStr` trait (ASCII, fullwidth characters, mixed text)

## Technical Details

### API Design

The trait-based approach provides:
- More idiomatic Rust API (`.terminal_width()` method calls)
- Consistent interface with existing `UnicodeWidthStr` trait
- Encapsulation of internal implementation details

### Background

In legacy Shift-JIS encoding (JIS X 0201), halfwidth katakana and dakuten/handakuten were completely separate 1-byte characters. Terminal emulators maintain this behavior for compatibility, rendering `ｶﾞ` as two distinct character cells rather than as a combined single character.

The combining variants (U+3099 and U+309A) properly combine with preceding characters and report correct width 0, so they don't need special handling.

### Known Design Issue and Future Improvements

This PR avoids breaking changes by keeping the existing `UnicodeWidthStr` trait implementations. However, there is a design inconsistency:

**Current Issue:**
- Types like `Span`, `Line`, `Text`, `Tabs` implement the `UnicodeWidthStr` trait
- However, their `.width()` methods actually return "terminal display width" instead of "Unicode standard width"
- The trait name doesn't match the actual behavior

**Future Improvements:**
- Remove `UnicodeWidthStr` implementations or change them to return true Unicode width
- Provide a custom `TerminalWidth` trait instead

For now, we maintain existing behavior to minimize user impact.

## References

- [Ruby reline PR #832](https://github.com/ruby/reline/pull/832) - Same issue and fix approach
- [Microsoft Terminal Issue #18087](https://github.com/microsoft/terminal/issues/18087) - Terminal behavior discussion
- [Unicode L2/19-039](https://www.unicode.org/L2/L2019/19039-grapheme-break.pdf) - Grapheme break property for U+FF9E and U+FF9F

g4yamanaka committed:
fix: correct terminal width for halfwidth katakana with dakuten

This commit fixes the terminal width calculation for halfwidth katakana
characters combined with dakuten (U+FF9E) and handakuten (U+FF9F).

# Problem

The halfwidth katakana dakuten and handakuten marks (U+FF9E and U+FF9F)
are non-combining marks that terminals display as separate characters,
each taking 1 column width. However, `unicode-width` reports them as
width 0 due to their `Grapheme_Extend` property in Unicode.

This discrepancy causes layout issues in terminal UIs when using these
characters, as ratatui was using unicode-width's values directly.

# Solution

Introduced `TerminalWidthStr` trait providing `.terminal_width()` method
that correctly handles halfwidth dakuten/handakuten marks by treating
them as width 1 characters, while using unicode-width for all other cases.

# Technical Details

- Added `terminal_width(grapheme: &str) -> usize` internal function
- Added `TerminalWidthStr` trait with `.terminal_width()` method for strings
- Updated all width calculations across the codebase to use the trait method
- Added comprehensive tests for various character combinations

# Background

In legacy Shift-JIS encoding (JIS X 0201), halfwidth katakana and
dakuten/handakuten were separate 1-byte characters. Terminal emulators
maintain this behavior for compatibility, rendering them as distinct
character cells rather than combining them.

g4yamanaka committed:
fix: clippy uninlined_format_args warnings

g4yamanaka committed:
style: run cargo fmt

g4yamanaka committed:
style: run cargo +nightly fmt

g4yamanaka committed:
chore: remove mise.toml

orhun commented:
Hey, thanks for investigating this issue and providing a fix!

Although the changes look good to me:

- I'm not sure if we go for another method of calculating widths (`TerminalWidthStr`). Right now most of the unicode width calculation is done by `unicode_width` crate and I'm afraid we might introduce inconsistency if we change/wrap that behavior.
- If this is an issue with width calculation and `unicode_width` is the root cause, wouldn't it be better to fix it there instead?
- What terminal did you use for testing this? For me Alacritty worked but Wezterm fails. Also it would be better to test against `0.30-beta` instead of `0.29` release.

See #1764 for a similar issue. Maybe this fix might apply there as well.

joshka commented:
If this is an issue with width calculation and `unicode-width` is the root cause, I think the crate is actually *doing the right thing* per the Unicode spec.

These characters (U+FF9E, U+FF9F) have the `Grapheme_Extend` property ([Unicode 17 §3 D59](https://www.unicode.org/versions/Unicode17.0.0/core-spec/chapter-3/#G41165)), and `unicode-width` correctly reports them as width 0. Terminals, however, intentionally render them as width 1 for Shift-JIS compatibility — so the mismatch lies between Unicode semantics and terminal conventions, not a bug in `unicode-width`.

Given that, I think Ratatui should explicitly commit to measuring **terminal display width**, not Unicode width. In other words, we should elevate `TerminalWidthStr` as the canonical abstraction for layout measurements and remove `UnicodeWidthStr` from our public API entirely.

To make this transition straightforward and explicit:

- Drop `unicode-width` from all public interfaces  
- Rename `terminal_width()` → `width()`  
- Remove existing `.width()` methods so downstream crates must update their imports from `UnicodeWidthStr` to `TerminalWidthStr`  

That way, the rename signals the intentional semantic shift, and the required import change keeps the migration mechanical and obvious.

Since `UnicodeWidthStr` implementations on `Span`/`Line`/`Text` only exist in the current beta, this would be a breaking change only for pre-release users — so we could do this before cutting 0.30.0 and publish a new beta. We don't need to mark this as a breaking change

---

**TL;DR**

- `unicode-width` is correct per Unicode, but terminals differ  
- Ratatui should standardize on *terminal width*, not *Unicode width*  
- Make `TerminalWidthStr` the golden path, rename `.terminal_width()` to `.width()`, and remove `UnicodeWidthStr` from the public surface  

How does that sound to you, @orhun?

---

*(As an aside, I recently worked on a [similar half-width character issue in Codex](https://github.com/openai/codex/pull/4359), which addressed this from the input side. That PR may not map directly here, but it shows the same class of discrepancies between input semantics and terminal rendering. @genki might find this interesting, though no obligation to weigh in.)*

joshka commented on ratatui-core/src/text/grapheme.rs:
````diff
@@ -47,6 +55,99 @@ impl Styled for StyledGrapheme<'_> {
+/// assert_eq!("ｶﾞｷﾞｸﾞ".terminal_width(), 6); // 3 graphemes × 2 width each
+/// assert_eq!("あいう".terminal_width(), 6); // 3 graphemes × 2 width each
+/// ```
+pub trait TerminalWidthStr {
````
Small nit: I think this and the above function belongs in a new file that's just about the width. Name it `width.rs`
That keeps the testing small and focused around this one aspect of things. There's not a lot of overlap on the grapheme code here.

joshka reviewed (changes requested):
This is a really good PR. Well tested, well explained. Just generally great.
See the wider notes in the PR comment I just made

TLDR
- Drop unicode-width from all public interfaces
- Rename terminal_width() → width()
- Remove existing .width() methods so downstream crates must update their imports from UnicodeWidthStr to TerminalWidthStr

joshka commented:
Hrm, actually I'm not sure now that you put it that way. Seems I forgot that there were width methods on the actual items. I think perhaps implementing UnicodeWidthStr was the wrong approach. Taking a look at what's there in 0.29 (https://docs.rs/ratatui/0.29.0/ratatui/index.html?search=width), maybe we could get away with instead using only the non-trait methods. WDYT?

g4yamanaka commented:
Background: I'm not an expert in this area, but I've investigated based on the review comments. Below are my proposals, though I'll defer to the maintainers' judgment
  on the final direction.

  Proposal 1: Remove UnicodeWidthStr
  - Remove impl UnicodeWidthStr for Span/Line/Text
  - Keep Span::width() / Line::width() / Text::width() methods as inherent methods
  - Don't migrate to trait

  Rationale:
  - Users no longer need to import external crate traits
  - Inherent methods are simpler and more intuitive

  Proposal 2: Remove width_cjk()
  - I propose removing the width_cjk() methods as well

  Context: What is width_cjk()?
  - width() = Calculates Ambiguous characters (box-drawing │, Greek α, etc.) as width 1
  - width_cjk() = Calculates Ambiguous characters as width 2
  - Ambiguous = Unicode East Asian Width property characters whose width "depends on context"
  - All box-drawing characters used by ratatui (│ U+2502, etc.) are Ambiguous
  - Complete list of Ambiguous characters: https://raw.githubusercontent.com/hamano/locale-eaw/master/test/amb.txt
  - Even in CJK locales (ja_JP.UTF-8, etc.), most terminal defaults treat Ambiguous as width 1
  - Requires explicit configuration to use width 2 (e.g., wezterm's treat_east_asian_ambiguous_width_as_wide = true)

  Reasons for removal (my opinion):

  1. Misleading expectations: The existence of width_cjk() might create false expectations that "ratatui supports CJK environments"
  2. Design limitation: ratatui calculates box-drawing widths as 1 in Block::inner(). Providing only width_cjk() won't work without layout system support
  3. Partial support concerns: Full Ambiguous=Wide support would require design changes across the entire system (Block::inner(), all widgets, rendering)

  Actual behavior verification:
  I tested ratatui's block example with wezterm:

  - Ambiguous=Narrow (default): 
    - Box-drawing characters display as width 1, layout works correctly
    - <img width="622" height="413" alt="narrow" src="https://github.com/user-attachments/assets/5ccca65b-cd4c-4cd4-b097-2f4e3f910936" />
  - Ambiguous=Wide (treat_east_asian_ambiguous_width_as_wide = true):
    - Box-drawing characters display as width 2, layout breaks
    - <img width="622" height="413" alt="wide" src="https://github.com/user-attachments/assets/525e45ed-55b6-43f5-a876-4c213e4c07a6" />


  This comparison shows that providing only width_cjk() may cause issues since the layout system doesn't support it.

  WDYT? @joshka @orhun

g4yamanaka commented:
@joshka @orhun

Just wanted to follow up on Proposal 1 & 2 above. Should I proceed with the implementation, or would you like to discuss further first?

Happy to wait for your confirmation before starting! 👍

g4yamanaka committed:
refactor: move TerminalWidthStr to dedicated width module

- Move TerminalWidthStr trait from grapheme.rs to new width.rs module
- Remove impl UnicodeWidthStr for Span, Line, Text, and Tabs
- Keep width() as inherent methods using TerminalWidthStr internally
- Remove width_cjk() methods (Proposal 2 from PR discussion)

g4yamanaka committed:
style: fix formatting and remove unused unicode-width dep

g4yamanaka committed:
refactor: replace UnicodeWidthStr with TerminalWidthStr in tests and examples

g4yamanaka committed:
ci: trigger CI rebuild

g4yamanaka committed:
refactor: make grapheme module private

a-kenji commented:
> What if we brute forced this by throwing a bunch of Unicode (literally all of it) at terminals (all of them), and detect which characters move the cursor by how much, then back fill this into into a bunch standard configs?

Yes, that sounds interesting. 
Also there are some escape sequences that can query the actual width of a certain character, if the terminal supports it. 

In which terminals can I reproduce the original problem? I tried running it in a couple terminals and I always get (ratatui v 30.0): 
<img width="540" height="366" alt="image" src="https://github.com/user-attachments/assets/92c6286d-c5eb-4853-a081-f0ad016df0b7" />

Even in the current default kitty version (kitty 0.44.0)
